libraries/AP_Frsky_Telem/mavlite.cpp

```diff
@@ -0,0 +1,288 @@
+            } else {
+                if ( status->payload_next_byte < rxmsg->len ) {
+                    rxmsg->payload[status->payload_next_byte++] = byte;
+                    mavlite_update_checksum(rxmsg, byte);
+                } else {
+                    if ( rxmsg->checksum == byte ) {
+                        status->parse_state = PARSE_STATE_MESSAGE_RECEIVED;
+                        return true;
+                    } else {
+                        status->parse_state = PARSE_STATE_ERROR;
+                    }
+                }
+            }
+            break;
+        case PARSE_STATE_MESSAGE_RECEIVED:
+            if ( offset == 0 && byte == 0x00 ) {
+                mavlite_init_parse(rxmsg, status);
+                status->parse_state = PARSE_STATE_GOT_START;
+                mavlite_update_checksum(rxmsg, byte);
+            }
+            break;
+    }
+    return false;
+}
+
+bool MAVLITE::mavlite_tx_parse(uint8_t* byte,uint8_t offset,  mavlite_message_t* txmsg, mavlite_status_t* status)
+{
+    switch (status->parse_state) {
+        case PARSE_STATE_IDLE:
+        case PARSE_STATE_ERROR:
+            if ( offset == 0 ) {
+                *byte = 0x00;
+                status->parse_state = PARSE_STATE_GOT_START;
+                mavlite_update_checksum(txmsg, *byte);
+            } else {
+                status->parse_state = PARSE_STATE_ERROR;
+            }
+            break;
+        case PARSE_STATE_GOT_START:
+            *byte = txmsg->len;
+            status->parse_state = PARSE_STATE_GOT_LEN;
+            mavlite_update_checksum(txmsg, *byte);
+            break;
+        case PARSE_STATE_GOT_LEN:
+            *byte = txmsg->msgid;
+            status->parse_state = PARSE_STATE_GOT_MSGID;
+            mavlite_update_checksum(txmsg, *byte);
+            break;
+        case PARSE_STATE_GOT_MSGID:
+            *byte = txmsg->payload[status->payload_next_byte++];
+            status->parse_state = PARSE_STATE_GOT_PAYLOAD;
+            mavlite_update_checksum(txmsg, *byte);
+            break;
+        case PARSE_STATE_GOT_SEQ:
+            if ( status->payload_next_byte < txmsg->len ) {
+                *byte = txmsg->payload[status->payload_next_byte++];
+                status->parse_state = PARSE_STATE_GOT_PAYLOAD;
+                mavlite_update_checksum(txmsg, *byte);
+            } else {
+                *byte = txmsg->checksum;
+                status->parse_state = PARSE_STATE_MESSAGE_RECEIVED;
+                return true;
+            }
+            break;
+        case PARSE_STATE_GOT_PAYLOAD:
+            if ( offset == 0) {
+                *byte = ++status->current_rx_seq;
+                mavlite_update_checksum(txmsg, *byte);
+                status->parse_state = PARSE_STATE_GOT_SEQ;
+            } else {
+                if ( status->payload_next_byte < txmsg->len ) {
+                    *byte = txmsg->payload[status->payload_next_byte++];
```

Comment:
Bounds-checking required.  This, however, would be an internal ArduPilot bug if it did happen.